### PR TITLE
Add Markdown round-trip support to AgentMemory

### LINKED ISSUE #765: Feature request: wiki-style editable memory format/view alongside graph and RAG memory
## Summary

Is there a roadmap for a wiki-style, human-editable memory format or memory view in Semantica?

After cross-checking the public docs, the documented memory/context surfaces appear to focus on:

- `ContextGraph` and graph-backed context
- `AgentMemory` and vector-backed retrieval
- GraphRAG / hybrid retrieval
- programmatic memory operations such as `store()`, `retrieve()`, `update()`, `forget()`, `export()`, and `import_data()`

That is useful for agent/runtime retrieval, but it is harder to manually curate, edit, diff, and reorganize memory in a straightforward human-readable way.

## Why this matters

For some workflows, teams want a memory layer that is easy for humans to:

- inspect directly
- edit or rewrite in place
- restructure without graph tooling
- review in diffs / version control
- maintain as a wiki-like knowledge base alongside agent memory

Right now, Semantica seems strong on graph and RAG-style memory, but less clear on a wiki-like memory UX.

## Request

Would you consider one of these?

1. A wiki/Markdown-style memory export that remains round-trippable back into Semantica.
2. A human-editable memory layer or explorer view for manual curation.
3. Guidance on the recommended pattern if this already exists but is not documented.

## Cross-check context

As of July 21, 2026, the docs I checked describe context graphs, agent memory, GraphRAG, and JSON/dict or graph-oriented export paths, but I did not find an obvious documented wiki-style memory format.

If there is already a supported approach, a pointer to the right docs would help.

Thanks.

### PR DESCRIPTION
**Pull request by SaurabhScripts**
## Summary

- add Markdown export and import to `AgentMemory`
- use one Markdown file per memory item with required YAML frontmatter
- preserve stable IDs, timestamps, supported metadata, entities, relationships, and exact body content
- create new memories or update matching IDs idempotently
- reject malformed, duplicate, or unsafe input before mutating memory
- keep Markdown import scoped to memory-local state without editing `ContextGraph`
- document the round-trip workflow and file contract

## Review fixes

- compare aware and local-naive timestamps safely across retention, recent sorting, and date filters
- reject symlink output targets and atomically replace Markdown files
- treat reused custom memory IDs as replacements so indexes and statistics remain consistent
- defer vector-store synchronization until local updates commit and track adapter vector IDs for safe replacement and deletion

## Why

`AgentMemory` currently exposes JSON/dict serialization, but no human-editable, diff-friendly representation for reviewing and curating memory. This adds a deterministic Markdown workflow while keeping Explorer and graph editing outside the MVP.

## Validation

- `python -m pytest -q tests/context/test_agent_memory_markdown.py` - 43 passed
- `python -m pytest -q tests/context` - 474 passed
- `python -m pytest -q tests/test_030_context_graph_realworld_extended.py -k AgentMemory` - 6 passed
- Black, isort, flake8, `git diff --check`, and Python 3.8 syntax checks passed

Closes #765

### CODE REVIEW

**Comment by KaifAhmad1** [comment]
@SaurabhScripts Thanks for the PR! Before I merge it, could you please address the Qodo review comments, especially the required timezone handling and symlink security fixes? If possible, please also address the vector store rollback and stats consistency recommendations. Thanks!

File: semantica/context/agent_memory.py [hunk 1]
```diff
@@ -1414,6 +1547,451 @@ def import_data(
...
+        if isinstance(value, datetime):
+            parsed = value
+        elif isinstance(value, date):
+            parsed = datetime.combine(value, datetime.min.time())
+        elif isinstance(value, str) and value.strip():
+            candidate = value.strip()
+            if candidate.endswith("Z"):
+                candidate = candidate[:-1] + "+00:00"
+            try:
+                parsed = datetime.fromisoformat(candidate)
```
[the remark this answers is not in the document]
Comment (Sameer6305) [reply semantica/context/agent_memory.py:2148]: Addressed in 40d6fa0.

_markdown_record_matches() now normalizes timestamps via _timestamp_comparison_key() before comparison, matching the comparison strategy used elsewhere in the implementation. Added a regression test covering naive/aware timestamp equivalence to preserve idempotent Markdown re-imports.

**Review by Sameer6305 — APPROVED** [review]
I performed a detailed review of the Markdown import/export implementation and the associated test suite.

### Review scope
- `semantica/context/agent_memory.py`
- `tests/context/test_agent_memory_markdown.py`
- `docs/reference/context.md`

### What I verified
- Markdown import/export round-trip behavior and deterministic serialization.
- YAML frontmatter parsing and duplicate-key validation.
- Stable memory ID, metadata, entity, and relationship preservation.
- Duplicate ID detection during import.
- Atomic file writes and filesystem safety checks.
- Rollback behavior during failed imports.
- Vector store synchronization after successful imports.
- Timestamp handling across import, export, retention, filtering, and recent-memory operations.
- Overall idempotency of repeated Markdown imports.
- Coverage provided by the accompanying test suite.

### Findings

During review, I found one remaining edge case in the idempotency check.

`_markdown_record_matches()` compared timestamps using direct equality, while the rest of the implementation already normalizes timestamps through `_timestamp_comparison_key()` for comparisons involving timezone-naive and timezone-aware datetimes. This could cause an unchanged Markdown import to perform an unnecessary replacement and vector synchronization when the timestamps represented the same instant but differed only in timezone representation.

### Resolution

I addressed the issue by:
- Normalizing timestamp comparison in `_markdown_record_matches()` using `_timestamp_comparison_key()`.
- Adding a focused regression test covering naive/aware timestamp equivalence to ensure idempotent Markdown re-import behavior.

The relevant Markdown test suite passes after the change.

Overall, the implementation is well structured, the import/export workflow is robust, and the feature is ready to merge.

**LGTM. Approving.** ✅

**Review by KaifAhmad1 — APPROVED** [review]
## Final Review — PR #786: Add Markdown round-trip support to `AgentMemory`

**Repo:** semantica-agi/semantica
**Author:** @SaurabhScripts
**Branch:** `codex/agent-memory-markdown-round-trip` → `main`
**Reviewers:** Qodo bot, Codex bot, @Sameer6305 (approved), @KaifAhmad1

---

### Overview

- Adds `format="markdown"` to `AgentMemory.export()` / `import_data()` as a human-editable, diff-friendly alternative to the existing JSON/dict serialization.
- One YAML-frontmatter Markdown file per memory item; required fields `id`, `created_at`, `updated_at`, and `type`/`kind`.
- Upsert-by-ID semantics: unknown IDs create new memories, known IDs are replaced atomically.
- Deferred, best-effort vector-store sync — local state commits first, vectors sync after.
- Entities/relationships round-trip as memory-local provenance only; `ContextGraph` is intentionally untouched, matching the MVP scope agreed on in the #765 issue thread.
- Docs updated in `docs/reference/context.md`; 43 new tests in `tests/context/test_agent_memory_markdown.py`.
- Stats: +1,544 / -53 across 3 files.

### Prior review rounds (verified, not just trusted)

- **Symlink safety (Qodo, resolved):** `_write_markdown_file` checks `is_symlink()`, writes to a temp file in the same directory, then `os.replace()`s — swaps the directory entry rather than following it. Confirmed correct.
- **Timezone-naive vs. aware comparison crash (Qodo, resolved):** `_timestamp_comparison_key()` is now used consistently across `_matches_filters`, `_apply_retention_policy`, `get_recent`, `get_by_date`, and `_markdown_record_matches`. No remaining raw `<`/`>`/`==` timestamp comparisons found.
- **Stats/index drift on reused custom IDs (Qodo, resolved):** `store()` now routes existing IDs through `_replace_memory_item`, which snapshots, deletes, re-stores, and rolls back on failure. `delete_memory` correctly decrements `items_by_type` and purges all matching entries from `memory_index`/`short_term_memory`.
- **Vector-store rollback gaps (Qodo, resolved):** local mutations use `skip_vector=True` and defer to `_sync_committed_vector_changes` after the in-memory state is committed, so a failed vector write can no longer trigger a local rollback that leaves stale vector IDs. `_vector_ids` persists through `save()`/`load()`.
- **Idempotency edge case (Sameer6305, resolved):** `_markdown_record_matches()` now normalizes timestamps via `_timestamp_comparison_key()` before comparing, so a naive/aware timestamp representing the same instant no longer triggers a spurious replace + vector sync on unchanged re-import.

### Additional findings from independent review (minor, non-blocking)

- **Error masking in path-detection heuristic** (`_import_markdown_payload`): the `try/except OSError: pass` around the "is this a path?" check wraps both `candidate.exists()` and the actual file read. A `PermissionError` while reading a real path is silently swallowed and the code falls through to treating the path string as raw Markdown text — producing a confusing "invalid frontmatter" error instead of surfacing the real permission problem. Suggest narrowing the try block or re-raising non-`FileNotFoundError` cases.
- **Read-side symlink asymmetry**: export refuses to write through a symlink, but `_read_markdown_path` will read through symlinked files/directories on import without restriction. Low risk (it's the user's own import directory) but inconsistent with the write-side hardening.
- **Path-vs-content ambiguity is by design, not a bug**: a single-line string with no newline is speculatively treated as a filesystem path. Safe in practice since valid frontmatter documents always span ≥3 lines, but the reasoning isn't documented inline.
- **Double snapshot/rollback on batch import**: `_import_markdown_records` snapshots once for the whole batch, but each `_replace_memory_item` call also snapshots/rolls back internally before re-raising. Correct, just redundant work on a mid-batch failure.

None of the above are correctness or security blockers.

### Test coverage

- Round-trip losslessness and deterministic re-export.
- Directory export/import with stable, content-hashed filenames.
- ID-reuse idempotency, including the tz-normalization edge case.
- Malformed/duplicate frontmatter and duplicate IDs rejected before any mutation.
- All-or-nothing directory import (partial failure rolls back the whole batch).
- Vector-store rollback and deferred-sync ordering, including a failing adapter.
- Symlink rejection on export.
- Legacy JSON/dict import/export behavior unchanged.

Coverage is thorough and matches the stated MVP scope; nothing obviously missing.

### Follow-up work done during this review cycle

- Reviewed the full diff and confirmed all four Qodo findings plus Sameer6305's idempotency fix were correctly implemented.
- Added a `CHANGELOG.md` entry under `[Unreleased] / Added` describing the feature, guarantees, and scope, pushed as a follow-up commit (`1c590c62`) to `SaurabhScripts:codex/agent-memory-markdown-round-trip` (maintainer-edit enabled on the PR).
- Resolved a `CHANGELOG.md` merge conflict against `main` (both branches had added a new `[Unreleased]` section independently) by keeping both entries with `### Added` ordered before `### Fixed`, per Keep a Changelog convention; merged and pushed (`87714ec1`).
- PR is now `mergeable: MERGEABLE` (was blocked by the conflict before).

### Verdict

**Ready to merge.** No correctness or security blockers. Already approved by @Sameer6305 with all required/recommended Qodo findings resolved; the two minor items noted above (error masking, read-side symlink asymmetry) are reasonable fast-follow polish, not merge gates.
